pkg/apis/stork/v1alpha1/types.go

```diff
@@ -308,3 +315,253 @@ const (
+}
+
+// CommonSpec are common configurations that are exposed at both
+// cluster and node level
+type CommonSpec struct {
+	// Network is the network information for storage driver
+	Network *NetworkSpec `json:"network"`
+	// Storage details of storage used by the driver
+	Storage *StorageSpec `json:"storage"`
+	// Env is a list of environment variables used by the driver
+	Env []corev1.EnvVar `json:"env"`
+	// RuntimeOpts is a map of options with extra configs for storage driver
+	RuntimeOpts map[string]string `json:"runtimeOptions"`
+}
+
+// NodeSelector let's the user select a node or group of nodes based on either
+// the NodeName or the node LabelSelector. If NodeName is specified then,
+// LabelSelector is ignored as that is more accurate, even though it does not
+// match any node names.
+type NodeSelector struct {
+	// NodeName is the name of Kubernetes node that it to be selected
+	NodeName string `json:"nodeName"`
+	// LabelSelector is label query over all the nodes in the cluster
+	LabelSelector *meta.LabelSelector `json:"labelSelector"`
+}
+
+// PlacementSpec has placement configuration for the storage cluster nodes
+type PlacementSpec struct {
+	// NodeAffinity describes node affinity scheduling rules for the pods
+	NodeAffinity *corev1.NodeAffinity `json:"nodeAffinity"`
+}
+
+// KvdbSpec contains the details to access kvdb
+type KvdbSpec struct {
+	// Internal flag indicates whether to use internal kvdb or an external one
+	Internal bool `json:"internal"`
+	// Endpoints to access the kvdb
+	Endpoints []string `json:"endpoints"`
+	// AuthSecret is name of the kubernetes secret containing information
+	// to authenticate with the kvdb. It could have the username/password
+	// for basic auth, certificate information or ACL token.
+	AuthSecret string `json:"authSecret"`
+}
+
+// NetworkSpec contains network information
+type NetworkSpec struct {
+	// DataInterface is the network interface used by driver for data traffic
+	DataInterface *string `json:"dataInterface"`
+	// MgmtInterface is the network interface used by driver for mgmt traffic
+	MgmtInterface *string `json:"mgmtInterface"`
+}
+
+// StorageSpec details of storage used by the driver
+type StorageSpec struct {
+	// UseAll use all available, unformatted, unpartioned devices.
+	// This will be ignored if Devices is not empty.
+	UseAll *bool `json:"useAll"`
+	// UseAllWithPartitions use all available unformatted devices
+	// including partitions. This will be ignored if Devices is not empty.
+	UseAllWithPartitions *bool `json:"useAllWithPartitions"`
+	// Devices list of devices to be used by storage driver
+	Devices *[]string `json:"devices"`
+	// JournalDevice device for journaling
+	JournalDevice *string `json:"journalDevice"`
+	// SystemMdDevice device that will be used to store system metadata
+	SystemMdDevice *string `json:"systemMetadataDevice"`
+	// DataStorageType backing store type for managing drives and pools
+	DataStorageType *string `json:"dataStorageType"`
```

Comment:
We should enums for Types

Reply:
These cannot be well defined enums if we have to keep the CRD provider agnostic. An example of data storage type would be `pxv`. Every storage provider can come have their custom data store types.